appinventor/blocklyeditor/src/blocks/components.js

```diff
@@ -197,6 +197,20 @@ Blockly.ComponentBlock.addGenericOption = function(block, options) {
   options.splice(options.length - 1, 0, item);
 };
 
+/**
+ * Marks the passed block as a badBlock() and disables it if the data associated
+ * with the block is not defined, or the data is marked as deprecated.
+ * @param {Blockly.BlockSvg} block The block to check for deprecation.
+ * @param {EventDescriptor|MethodDescriptor|PropertyDescriptor} data The data
+ *     associated with the block which is possibly deprecated.
+ */
+Blockly.ComponentBlock.checkDeprecated = function(block, data) {
+  if ((!data || data.deprecated) && block.workspace == Blockly.mainWorkspace) {
+    block.badBlock();
```

Comment:
Setting the block to disabled overrides the visible effect of `badBlock`, so you could eliminate this call.

Reply:
Yeah I thought it was weird that the behavior was to disabled blocks. Will fix :D

Reply:
I think we are mixing two different issues here. In the case where `data` is undefined, the block definition no longer exists. This is a truly `badBlock` and can't generate code because we don't have enough information to do so. In the case where the item is marked deprecated, the underlying code still exists and so we could disable the block (re-enabling will restore the old behavior) with the expectation that some day the block will go away. Alternatively, we could include an error message stating that the block is deprecated and should be removed. This mainly applies to extensions as if we do remove a block we usually have an upgrader to substitute in the replacement. Extensions don't have upgraders though so can't provide similar support.

Reply:
Ok I changed it so that if the functions `@Deprecated` the blocks are disabled. If the functions are removed, the blocks are marked as `badBlock`s.

So if you **deprecate** the functions for the blocks in the test aia, they look like this:
![Deprecation](https://user-images.githubusercontent.com/25440652/86038715-3590d900-b9f6-11ea-949e-7505cb4feec5.png)

If you **remove** the functions, they look like this:
![removed-blocks](https://user-images.githubusercontent.com/25440652/106072519-31788700-60bd-11eb-98b7-6b2fd62c98de.png)